cmd/query/app/apiv3/grpc_gateway_test.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2021 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apiv3
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+
+	"github.com/jaegertracing/jaeger/cmd/query/app/querysvc"
+	"github.com/jaegertracing/jaeger/model"
+	_ "github.com/jaegertracing/jaeger/pkg/gogocodec" //force gogo codec registration
+	"github.com/jaegertracing/jaeger/proto-gen/api_v3"
+	dependencyStoreMocks "github.com/jaegertracing/jaeger/storage/dependencystore/mocks"
+	spanstoremocks "github.com/jaegertracing/jaeger/storage/spanstore/mocks"
+)
+
+func TestGRPCGateway(t *testing.T) {
+	r := &spanstoremocks.Reader{}
+	traceID := model.NewTraceID(150, 160)
+	r.On("GetTrace", mock.AnythingOfType("*context.valueCtx"), mock.AnythingOfType("model.TraceID")).Return(
+		&model.Trace{
+			Spans: []*model.Span{
+				{
+					TraceID:       traceID,
+					SpanID:        model.NewSpanID(180),
+					OperationName: "foobar",
+				},
+			},
+		}, nil).Once()
+
+	q := querysvc.NewQueryService(r, &dependencyStoreMocks.Reader{}, querysvc.QueryServiceOptions{})
+	server := grpc.NewServer()
+	h := &Handler{
+		QueryService: q,
+	}
+	api_v3.RegisterQueryServiceServer(server, h)
+
+	lis, _ := net.Listen("tcp", ":0")
+	go func() {
+		err := server.Serve(lis)
```

Comment:
Will this not leak the server & goroutine after the test is done?  (see https://github.com/jaegertracing/jaeger/pull/2606)

Same Q for http server. I would prefer the tests to exit cleanly & shut down the servers.

Reply:
there was a leak. it is fixed now